cosima_cookbook/database.py

```diff
@@ -600,16 +568,60 @@ def build_index(
         directories = [directories]
 
     indexed = 0
-    for directory in directories:
-        indexed += index_experiment(
-            directory, session, client, update, prune, delete, followsymlinks
-        )
+    for directory in [Path(d) for d in directories]:
+        # find all netCDF files in the experiment directory
+        files = find_files(directory, followsymlinks=followsymlinks)
+        expt = find_experiment(session, directory)
+
+        if expt is None:
+            expt = NCExperiment(
+                experiment=str(directory.name), root_dir=str(directory.resolve())
+            )
+        else:
+            if not update:
+                print(
+                    "Not re-indexing experiment: {}\n"
+                    "Pass `update=True` to build_index()".format(directory.name)
+                )
+                continue
+
+        print("Indexing experiment: {}".format(directory.name))
+
+        if prune:
+            # Prune files that exist in the database but are not present
+            # on the filesystem
+            missing_files = set([f.ncfile for f in expt.ncfiles]) - files
+            if len(missing_files) > 0:
+                # Make a list of NCFile objects to pass to prune_files
+                missing_files_objs = [
```

Comment:
@angus-g Not really happy with this. Would prefer to just pass a list of files, given the `expt` object is available in `prune_files` it should be simple to recreate a list of `NCFile` objects but AFAICT there is no dedicated way to do that lookup. If there is a simple addition that would allow this I'd probably prefer doing that lookup within `prune_files`.

Reply:
I thought we were going to pull pruning out of indexing and just leave it up to `prune_experiment`? Maybe the issue still remains, just relocated. One solution could be to make an `ncfile_paths` association proxy on `NCExperiment`, which might also get some benefits of laziness or better pushing some of this operation into the DB layer...

Reply:
I think there needs to pruning done when indexing, or you risk leaving the DB in a "broken" state. No?

I agree all I've really done is push the prune logic up one layer in the code, but it feels like a better place for it if it is still required here.